Refine Throughput Test to add test scenarios and throughput targets based on current implementation

Ayiga opened the pull request:
<!-- Closes #<ISSUE_NUMBER> -->
<!-- These comments should help create a useful PR message, please delete any remaining comments before opening the PR. -->
<!-- If there is no issue number make sure to describe clearly *why* this PR is necessary. -->
<!-- Mention open questions, remaining TODOs, if any -->

### This PR:
<!-- Describe what this PR adds to this repo and why -->
<!-- E.g. -->
<!-- * Implements feature 1 -->
<!-- * Fixes bug 3 -->
Modifies the `TestTransactionStreamerEspressoThroughput` test to have specific test cases in four scenarios.  Throughput targets are set based on the observed values for the `Molten` integration.

The four scenarios are a consequence of testing throughput of standard progress 4 Messages/s at 3 KiB per Message, or catchup, being 1,000 messages behind at 3 KiB per Message.

For these two scenarios the throughput is being measured with an Idealized Builder, and a Max Block Size Builder to compare against real world observed restrictions versus the idealized version.

As a consequence, message generation has been adjusted in an effort to make this easier to test against.

<!-- ### This PR does not: -->
<!-- Describe what is out of scope for this PR, if applicable. Leave this section blank if it's not applicable -->
<!-- This section helps avoid the reviewer having to needlessly point out missing parts -->
<!-- * Implement feature 3 because that feature is blocked by Issue 4 -->
<!-- * Implement xyz because that is tracked in issue #123. -->
<!-- * Address xzy for which I opened issue #456 -->

<!-- ### Key places to review: -->
<!-- Describe key places for reviewers to pay close attention to -->
<!-- * file.rs, `add_integers` function -->
<!-- Or directly comment on those files/lines to make it easier for the reviewers -->

<!-- ### How to test this PR:  -->
<!-- Optional, uncomment the above line if this is relevant to your PR -->
<!-- If your PR is fully tested through CI there is no need to add this section -->
<!-- * E.g. `just test` -->

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->
<!-- [ ] For integration projects, make sure it won't break backward compatibility. -->

<!-- Details on maintaining backward compatibility for integration projects: -->
<!-- * Try to keep changes additive: add new, optional methods, flags, or parameters instead of modifying or removing existing functionality. -->
<!-- * If modification is necessary, it should be either a clear bug fix or guarded by a config/feature flag.  -->
<!-- * Follow Open Closed Principle and Interface Segregation Principle, clients should not be forced to depend on interfaces they do not use. -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210850486591816
  - https://app.asana.com/0/0/1210850486591820

Ayiga committed:
Add BlockBuilder for MockEspressoChain

In order to simulate real-world conditions it is beneficial to be able to simulate a
non-ideal block production that we currently see on the Espresso Chain.  The Blocks
that Espresso produces are essentially size constrained in order to ensure that the
production process doesn't tax the system more than it needs to be.

These restrictions can have an impact on the throughput rates we are able to
achieve.

As a result, it is beneficial to be able to simulate these restrictions in order to more
closely compare oujr tests against real-world scenarios.

Ayiga committed:
Revise TestTransactionStreamerEspressoThroughput conditions

The TestTransactionStreamerEspressoThroughput test originally was created just to
monitor throughput based on the parameters being configured for the mock
setups.  The idea was to determine how interval configuration impacted the integration
performance under idealized conditions.

With some base lines set, it is now important to try and ensure that our integration
continues to perform at an established baseline and well beyond in order to ensure
that we ware able to catch up in a sufficient amount of time, should we fall behind.

To achieve this goal the TestTransactionStreamerEspressoThroughput test has been
reworked to test specific scenarios against the required benchmarks.  Additionally
it is important to codify the requirements of it being able to outperform the baseline
by a factor.  In this case the factor chosen is around 8x, but this was chosen arbitrarily
and our tests have indicated that we are able to exceed this with the right polling
interval configuration.

Sneh1999 commented on system_tests/espresso/chain/block_builder.go:
```diff
@@ -0,0 +1,155 @@
+	return txs, nil
+}
+
+// MaxSizeRestrictedBuilder is a BlockBuilder that restricts the size combined
```
I am bit worried about this approach because lets say we want a builder which has max size restriction and also max throughput restriction. Then do we need to like creating another struct?

Why can we have a builder where we can enable/disable features so that we dont have to create a different builder for each feature? Any thoughts?

Ayiga replied:
I think the most important aspect is to ensure that we have concrete abstractions.  Actual implementation can vary and even be pluggable with wrapped implementations that influence behavior as needed.

In general, I don't think enabling things via `boolean`s or other feature flags should translate directly into the underlying structure, as evaluating a condition should only really be done at the time of object creation, and not during the runtime time of the task / process.  This is a performance consideration, which when we're talking about throughput is largely important.  If you can avoid a conditional branch from occurring every time the runtime task is evaluated, and instead incur that same conditional check at creation time then the underlying logic behaves consistently, and there's no need to second guess what's happening (It should reduce cognitive burden in most cases).

Consider the following:

```go
type Doer interface {
    Do() Done
}

func CreateDoer(doBar bool) Dooer {
    return &Foo{
        doBar: doBar
    }
}

func (foo *Foo) Do() Done {
    if foo.doBar {
        return foo.doBar()
    }
 
     return foo.doFoo()   
}
```

In this case every time `Do` is invoked, the conditional branch is evaluated, and we'll potentially go down a different conditional path.  If we consider the idea that this flag isn't really every mutated during the lifetime of this object, then why do we need to evaluate this condition and potentially incur any stack unwinding penalty for a branch prediction failure?

Instead, it seems that this could instead be done at creation time with two different implementations to choose between.

```go
type Doer interface {
    Do() Done
}


func CreateDoer(doBar bool) Doer {
    if (doBar) {
        return new(Bar)
    }
    return new(Foo)
}
```

Now the underlying implementations are theirs to control.

I can see a scenario against this which makes sense, and that is having the behavior change mid-runtime.  Say you suddenly decide you want the behavior to change after some event.  In the first approach you can just mutate the underlying object to change the flag.  However, this can also be done with the specific object, but instead of mutating the underlying object, you replace it entirely.  Thereby you still only incur the evaluation of conditional one extra time, which is substantially better for readability, cognitive load, maintainability, and performance.

As a general rule, I tend to favor objects that aren't mutated where possible.  This isn't always possible, sadly, but where it can be achieved it can lead to better cleaner solutions (At least in my opinion).

Sneh1999 commented on system_tests/espresso/espresso_transaction_streamer_throughput_test.go:
```diff
@@ -19,6 +19,84 @@ import (
+func RunChains(ctx context.Context, t *testing.T, mockEspressoChain *chain.MockEspressoChain, streamer *arbnode.TransactionStreamer, behaviors ...generate.GenerationBehavior) {
+	messagesInChannel := make(chan generate.Message, 10)
+	// Produce Espresso Blocks at a 2 second interval
+	go chain.ProduceEspressoBlocksAtInterval(ctx, mockEspressoChain, 2*time.Second)
```
nit: should we make this 2 seconds configurable?

Ayiga replied:
At the moment this is just a helper function for this single tests.  No case in this test yet requires this level of configuration.  Though I suppose this helper function doesn't really need to exist in this form at this point.

To be fair I had originally written it when I didn't have the `scenario` struct simplifying the creation and evaluation of the test cases.

Sneh1999 commented on system_tests/espresso/generate/messages.go:
```diff
@@ -132,102 +73,132 @@ func (g simpleGenerator) GenerateMessage(
-					have, want,
-				))
+			if err := sendMessageToWriter(w, m); err != nil {
+				return err // Return any error encountered while sending the message
```
`Return any error encountered while sending the message` this comment is redundant because reader can easily understand what is happening here

Ayiga replied:
Removed in 04443f31183cc93fda32949360fcc26a5286aa93.

Sneh1999 commented on system_tests/espresso/generate/messages.go:
```diff
@@ -132,102 +73,132 @@ func (g simpleGenerator) GenerateMessage(
+		}
+
+		// perform the next behavior with the given index
+		i = behavior(ctx, hasher, ch, i)
```
redundant comment: `// perform the next behavior with the given index`, code itself is self explanatory

Ayiga replied:
Removed in 04443f31183cc93fda32949360fcc26a5286aa93.

Sneh1999 commented on system_tests/espresso/generate/messages.go:
```diff
@@ -132,102 +73,132 @@ func (g simpleGenerator) GenerateMessage(
-		// Use the maximum value for MessageIndex
-		0xFFFFFFFFFFFFFFFF,
-	)
+// This sized is based on the average size of the messages observed in Molten
```
```
// DefaultMoltenMessageSize is the default size of the messages generated for
// Molten environments.
```

and 
```
This sized is based on the average size of the messages observed in Molten
```

these two comments explain the same things, maybe they can be combined?

Ayiga replied:
I'm confused by this suggestion.  These two statements are already within the same comment block?

I suppose you're saying that the information convey within the two lines is too similar.  I think he two statements are saying two different things.  The first line is stating the purpose of the of the variable and what it defines, and the second line is stating how we arrived at this apparent magic number justifying why we have defined it as such.

I think the only overlap is that they both make mention of `Molten environments`.  This is perhaps redundant.  To be fair, as well, the number comes from Zach based on his observations of the messages flowing through Molten, and not directly on individual observations of Molten.

Sneh1999 commented on system_tests/espresso/espresso_transaction_streamer_throughput_test.go:
```diff
@@ -30,93 +108,237 @@ func TestTransactionStreamerEspressoThroughput(t *testing.T) {
+			}
+
+			if have, want := float64(len(receivedMessages))/timingData.Duration.Seconds(), scenario.messageThroughput; !want.IsGreaterThanOrEqualTo(have) {
+				t.Errorf("message throughput is not within expected range:\nhave:\n\t%.2f messages/s\nwant:\n\t>= %s messages/s\n",
```
should we also use FatalF instead of ErrorF here similar to all other checks in the file?

Ayiga replied:
The difference between `Errorf` and `Fatalf` is that an `Errorf` allows the test to continue whereas `Fatalf` stops the rest of the test from running.

If we utilized `Fatalf` we would stop after the first failure.  By utilizing `Errorf` it allows multiple failure results to be identified and seen within the same test at once.  This allows for the identification of potentially multiple failures in a single test run.  This can allow for faster iteration as you can resolve potentially multiple issues in one test failure rather than having to fix and re-run the test for each failure identified one at a time.

As a general rule, you should utilize `Errorf` wherever you can, as it allows you to identify failure conditions.  You should **only** utilize `Fatalf` if the rest of the test cannot continue as the result of the failure.

Sneh1999 commented on system_tests/espresso/espresso_transaction_streamer_throughput_test.go:
```diff
@@ -30,93 +108,237 @@ func TestTransactionStreamerEspressoThroughput(t *testing.T) {
+			}
+
+			if have, want := float64(totalBytesProcessed)/timingData.Duration.Seconds(), scenario.sizeThroughput; !want.IsGreaterThanOrEqualTo(have) {
+				t.Errorf("size throughput is not within expected range:\nhave:\n\t%.2f bytes/s\nwant:\n\t>= %s bytes/s\n",
```
same here

Sneh1999 reviewed (commented):
Overall looks great! Just added a few comments

Ayiga committed:
Remove redundant comments

Based on feedback provided by @sneh1999, comments that just state what is being
explicitly performed have been removed as they don't add anything meaningful to
what has been written / defined.

- https://github.com/EspressoSystems/nitro-espresso-integration/pull/692#discussion_r2247997061
- https://github.com/EspressoSystems/nitro-espresso-integration/pull/692#discussion_r2247999152

These comments are just stating what's already visibly being done.  They don't need
to exist in any meaningful way.   This commit removes them as per the suggestion.

Sneh1999 reviewed (approved):
Thanks @Ayiga! Appreciate all the work here

We should def create a follow up ticket to document all this in a README but that can be done as backlog task!

zacshowa commented on system_tests/espresso/generate/messages.go:
```diff
@@ -132,102 +73,131 @@ func (g simpleGenerator) GenerateMessage(
+// messages of a specific size at a specified interval. This is useful for
+// simulating a steady stream of messages being produced over time at
+// a consistent interval and time.
+func ProduceMessagesOfSizeAtInterval(n arbutil.MessageIndex, size uint64, interval time.Duration) GenerationBehavior {
```
Would it be potentially beneficial to add a `GenerationBehavior` that can add some random [jitter](https://en.wikipedia.org/wiki/Jitter) to the timing of messages in a future PR? 
I think it would be easy to implement, e.g.
```
case <- ticker.C:
  sleeptime = rand(range)
  sleep(sleeptime)
  ch <- GenerateMessage(i+j, hasher, size)
```

If this doesn't sound useful feel free to ignore.

Ayiga replied:
I do think this would be beneficial, if nothing else just to test some variance.

zacshowa commented on system_tests/espresso/generate/messages.go:
```diff
@@ -132,102 +73,131 @@ func (g simpleGenerator) GenerateMessage(
+// GenerateStandardMoltenMessages is a GenerationBehavior that generates N
+// messages with the specific observed Molten standards.
+func GenerateStandardMoltenMessages(n arbutil.MessageIndex) GenerationBehavior {
+	return ProduceMessagesOfSizeAtInterval(n, DefaultMoltenMessageSize, 250*time.Millisecond)
```
```suggestion
	return ProduceMessagesOfSizeAtInterval(n, DefaultMoltenMessageSize, DefaultMoltenTimeBetweenMessages)
```

zacshowa reviewed (approved):
I have a couple small nits, but otherwise this LGTM. 
I think it would be beneficial to have some additional documentation on this in the future like Sneh suggested.

Ayiga committed:
Rename Throughput test

In order to ensure that the Throughput is picked up by ci, it's beneficial to have its
name follow the existing `run-espresso-ci-job` scheme.  That is that it's name
should contain `TestEspresso`.  This alone won't ensure that it's evaluated in ci, but
it helps to be consistent in order to identify these tests more effectively.

Ayiga committed:
Modify tests being targeted by run-espresso-ci-job

The `run-espresso-ci-job` will only target tests that contain the name `TestEspresso`
in them, and are contained directly within the `system_tests` directory.  It would be
beneficial to target any `TestEspresso` tests that are a descendent of the `system_tests`
instead of just residing directly within it.

This change expands the tests being considered to be any descendents of `system_tests`
rather than just sitting in that single folder.
